llama_cpp/llama_chat_format.py

```diff
@@ -1686,9 +1687,731 @@ def message_to_str(msg: llama_types.ChatCompletionRequestMessage):
+    def generate_streaming_tool_call(self, prompt: str):
+        # Generate function name
+        function_name = ""
+        stops = ["\n"]  # "{self.recipient_token}{func_name}\n{self.content_token}"
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=None)
+        for chunk in completion:
+            function_name += chunk["choices"][0]["text"]
+        function_name = function_name.strip()
+        prompt += function_name + f"\n{self.content_token}"
+        yield {"name": function_name, "arguments": ""}, None, None, chunk, prompt
+        # Generate function args
+        stops = [self.stop_token, f"\n{self.from_token}", "\n\n\n"]
+        grammar = self.get_grammar(function_name) if function_name != "python" else None
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=grammar)
+        for chunk in completion:
+            delta_text = chunk["choices"][0]["text"]
+            if delta_text:
+                prompt += delta_text
+                yield {"name": None, "arguments": delta_text}, None, None, chunk, prompt
+
+class FunctionaryV25ChatHandler(FunctionaryChatHandler):
+    tool_call_token = "<|reserved_special_token_249|>"
+    stop_token = "<|eot_id|>"
+    
+    def prepare_for_generation_with_tool_func_choice(self, prompt: str, tool_name: str):
+        prompt += f"{self.tool_call_token}{tool_name}\n"
+        stops = [self.tool_call_token, self.stop_token]
+        grammar = self.get_grammar(tool_name)
+        
+        return prompt, stops, grammar
+    
+    def generate_to_determine_text_or_tool_call(
+        self, prompt: str, tool_func_choice: Optional[Union[str, Dict]], completion_tokens: int
+    ):
+        # Generate 1 token first to determine text or tool call
+        stops = []
+        grammar = None
+        completion = self.create_completion(
+            prompt=prompt, stop=stops, grammar=grammar, max_tokens=1
+        )
+        completion_text = completion["choices"][0]["text"]
+        completion_tokens += completion["usage"]["completion_tokens"]
+        
+        prompt += completion_text
+        
+        if completion_text.strip() == self.tool_call_token:
+            return prompt, completion_tokens, "tool_call"
+        elif len(completion_text) == 0:
+            return prompt, completion_tokens, "stop"
+        else:
+            return prompt, completion_tokens, "content"
+        
+    def generate_content(self, prompt: str, content: str, completion_tokens):
+        stops = [self.tool_call_token, self.stop_token]
+        completion = self.create_completion(prompt=prompt, stop=stops, grammar=None)
+        completion_text = completion["choices"][0]["text"]
+        completion_tokens += completion["usage"]["completion_tokens"]
+        prompt += completion_text
+        content_prefix = "<|start_header_id|>assistant<|end_header_id|>\n\n"
+        content += prompt[prompt.rindex(content_prefix) + len(content_prefix):] + completion_text
```

Comment:
When testing this out I see responses like `The smart curtains have been closed. smart curtains have been closed.`.

I am not sure, but from what I can see here this first appends the `completion_text` to the `prompt`, then gets the tail of the `prompt` and adds `completion_text` again to the `content`.  Does this mean `completion_text` was added twice?

Reply:
Applying this fix *seems* to address the issue:

```suggestion
        content += prompt[prompt.rindex(content_prefix) + len(content_prefix):]
```